File: src/AgentOrchestrator-python/sdk_labs/code_along.py
"""AI Genius S5E2 — LIVE CODE-ALONG (starter).

Run one phase at a time::

    uv run python -m sdk_labs codealong --phase 1

Only the Phase 3 block marked TYPE THIS LIVE is edited on stage.
Fall back to ``code_along_final`` at any time.
"""

from typing import Annotated

from copilot import (
    CopilotClient,
    PermissionHandler,
    SessionEvent,
    SessionEventType,
    ToolInvocation,
    define_tool,
)
from pydantic import BaseModel, Field

from sdk_labs import model_picker
from sdk_labs._common import IdleWaiter, flatten, send_and_print

# Phase 0 — in-memory demo data
TRANSACTIONS: list[tuple[str, float, str]] = [
    ("C001", 245.50, "Grocery"),
    ("C001", 89.99, "Electronics"),
    ("C002", 32.00, "Grocery"),
    ("C002", 15.50, "Health"),
    ("C003", 1250.00, "Electronics"),
    ("C003", 450.00, "Fashion"),
    ("C004", 12.99, "Grocery"),
    ("C004", 8.50, "Grocery"),
    ("C005", 675.00, "Electronics"),
    ("C005", 320.00, "Fashion"),
]


# Phase 1 — Hello World: a client and a session
async def phase1_hello_world(requested_model_id: str | None) -> int:
    print("== Phase 1: hello world ==\n")

    async with CopilotClient() as client:
        model_id = await model_picker.pick(client, requested_model_id)
        if model_id is None:
            return 1

        # Streaming emits token deltas as they arrive.
        session = await client.create_session(model=model_id, streaming=True)

        async with session:
            waiter = IdleWaiter()

            def on_event(evt: SessionEvent) -> None:
                if evt.type is SessionEventType.ASSISTANT_MESSAGE_DELTA:
                    print(evt.data.delta_content or "", end="", flush=True)
                elif evt.type is SessionEventType.SESSION_ERROR:
                    print(f"\nERROR: {evt.data.message}")
                waiter.handle(evt)

            session.on(on_event)

            print("Prompt: Name three retail KPIs. One line each.\n")
            await session.send("Name three retail KPIs. One line each.")
            await waiter.wait()
            print()

    return 0


# Phase 2 — Events: what the session emits
async def phase2_events(requested_model_id: str | None) -> int:
    print("== Phase 2: events ==\n")

    async with CopilotClient() as client:
        model_id = await model_picker.pick(client, requested_model_id)
        if model_id is None:
            return 1

        session = await client.create_session(model=model_id, streaming=True)

        async with session:
            waiter = IdleWaiter()
            counters = {"order": 0, "deltas": 0}
            finished = False

            # Count streaming deltas so the lifecycle remains readable.
            def on_event(evt: SessionEvent) -> None:
                nonlocal finished
                name = evt.type.value

                # Keep shutdown events from printing after the summary.
                if finished:
                    return

                if "delta" in name.lower():
                    counters["deltas"] += 1
                    return

                counters["order"] += 1
                print(f"{counters['order']:3d}. {name}")

                if evt.type is SessionEventType.SESSION_IDLE:
                    finished = True

                waiter.handle(evt)

            session.on(on_event)

            await session.send("In one sentence: what is customer churn?")
            await waiter.wait()
            print(
                f"\n(+ {counters['deltas']} delta events suppressed"
                " — that flood IS the streaming)"
            )

    return 0


# Phase 3 — Tools: descriptions tell the model when to call the tool
class GetCustomerTotalParams(BaseModel):
    """Parameter schema sent to the model."""

    customer_id: Annotated[str, Field(description="Customer identifier, for example C003")]


@define_tool(description="Gets the total amount a given retail customer has spent.")
def get_customer_total(params: GetCustomerTotalParams, _invocation: ToolInvocation) -> str:
    matches = [t for t in TRANSACTIONS if t[0].casefold() == params.customer_id.casefold()]

    if not matches:
        return f"No transactions found for {params.customer_id}."

    total = sum(t[1] for t in matches)

    # Visible proof that the custom tool executed.
    print(f"  [tool] get_customer_total({params.customer_id}) -> ${total:,.2f}")
    return f"{params.customer_id} has {len(matches)} transactions totalling ${total:,.2f}."


async def phase3_tools(requested_model_id: str | None) -> int:
    print("== Phase 3: tools ==\n")

    async with CopilotClient() as client:
        model_id = await model_picker.pick(client, requested_model_id)
        if model_id is None:
            return 1

        # Run once with tools=[] to demonstrate the CLI's built-in tools.
        # TYPE THIS LIVE (Phase 3 only): replace it with tools=[get_customer_total].
        # Python tool calls require on_permission_request; approve_all is silent.
        # tools adds custom tools but does not remove built-ins.
        session = await client.create_session(
            model=model_id,
            streaming=False,
            tools=[],
            on_permission_request=PermissionHandler.approve_all,
        )

        async with session:
            waiter = IdleWaiter()

            def on_event(evt: SessionEvent) -> None:
                if evt.type is SessionEventType.ASSISTANT_MESSAGE:
                    # Tool-request messages have no displayable content.
                    if evt.data.content:
                        print(f"\nAssistant: {flatten(evt.data.content)}")
                elif evt.type is SessionEventType.SESSION_ERROR:
                    print(f"\nERROR: {evt.data.message}")
                waiter.handle(evt)

            session.on(on_event)

            # Let the model select the tool without naming it in the prompt.
            print("Prompt: How much has customer C003 spent in total?\n")
            await session.send("How much has customer C003 spent in total?")
            await waiter.wait()

    return 0


# Phase 4 — Persistence: close, then resume
async def phase4_persistence(requested_model_id: str | None) -> int:
    import uuid

    print("== Phase 4: persistence ==\n")

    async with CopilotClient() as client:
        model_id = await model_picker.pick(client, requested_model_id)
        if model_id is None:
            return 1

        # The session ID enables resume after closing.
        session_id = f"codealong-{uuid.uuid4().hex}"[:24]
        print(f"Session id: {session_id}\n")

        print("--- Turn 1 (new session) ---")
        session = await client.create_session(
            session_id=session_id, model=model_id, streaming=False
        )
        async with session:
            await send_and_print(
                session,
                "Remember this: my favourite retail segment is 'At Risk'. Reply with just OK.",
            )

        print("\nSession closed.\n")

        print("--- Turn 2 (resumed) ---")
        resumed = await client.resume_session(session_id, model=model_id, streaming=False)
        async with resumed:
            await send_and_print(resumed, "Which retail segment did I say was my favourite?")
    return 0


# Phase 5 — MCP: attach tools from a server
# Network dependent: use the recording if learn.microsoft.com is blocked.
async def phase5_mcp(requested_model_id: str | None) -> int:
    print("== Phase 5: MCP ==\n")

    async with CopilotClient() as client:
        model_id = await model_picker.pick(client, requested_model_id)
        if model_id is None:
            return 1

        session = await client.create_session(
            model=model_id,
            streaming=False,
            mcp_servers={
                "microsoft.docs.mcp": {
                    "type": "http",
                    "url": "https://learn.microsoft.com/api/mcp",
                    # Python requires this to expose the MCP server's tools.
                    "tools": ["*"],
                }
            },
            on_permission_request=PermissionHandler.approve_all,
        )

        async with session:
            waiter = IdleWaiter()

            def on_event(evt: SessionEvent) -> None:
                if evt.type is SessionEventType.TOOL_EXECUTION_START:
                    # Show whether each call came from MCP or a built-in tool.
                    server = getattr(evt.data, "mcp_server_name", None)
                    if server:
                        print(f"  [mcp] {server} :: {evt.data.mcp_tool_name}")
                    else:
                        print(f"  [built-in] {getattr(evt.data, 'tool_name', '?')}")
                elif evt.type is SessionEventType.ASSISTANT_MESSAGE:
                    if evt.data.content:
                        print(f"\nAssistant: {flatten(evt.data.content)}")
                elif evt.type is SessionEventType.SESSION_ERROR:
                    print(f"\nERROR: {evt.data.message}")
                waiter.handle(evt)

            session.on(on_event)

            # Explicit steering prevents a built-in web fetch from hiding the MCP call.
            print("Prompt: Search Microsoft Learn — what is Azure Container Apps?\n")
            await session.send(
                "Use the microsoft.docs.mcp tools to search Microsoft Learn, then tell me "
                "in two sentences what Azure Container Apps is."
            )
            await waiter.wait()

    return 0


# Dispatch
PHASES = {
    1: phase1_hello_world,
    2: phase2_events,
    3: phase3_tools,
    4: phase4_persistence,
    5: phase5_mcp,
}


async def run(phase: int, requested_model_id: str | None) -> int:
    handler = PHASES.get(phase)
    if handler is None:
        print(
            "Usage:\n"
            "  uv run python -m sdk_labs codealong --phase <1-5>\n\n"
            "  1  Hello world     — client + session, streaming\n"
            "  2  Events          — what the session emits, in order\n"
            "  3  Tools           ⭐ define a tool the model calls\n"
            "  4  Persistence     — close, then resume the same session\n"
            "  5  MCP             — attach a server you didn't write (needs network)"
        )
        return 1
    return await handler(requested_model_id)

File: src/AgentOrchestrator/samples/SdkLabs/CodeAlong.cs
using System.ComponentModel;
using GitHub.Copilot;

namespace SdkLabs;

/// <summary>
/// AI Genius S5E2 — LIVE CODE-ALONG (starter)
///
/// Run one phase at a time:
///     dotnet run --project src/AgentOrchestrator/samples/SdkLabs -- codealong --phase 1
///
/// Only the Phase 3 block marked TYPE THIS LIVE is edited on stage.
/// Fall back to CodeAlongFinal at any time.
/// </summary>
public static class CodeAlong
{
    // Phase 0 — in-memory demo data
    private static readonly (string CustomerId, decimal Amount, string Category)[] Transactions =
    [
        ("C001", 245.50m, "Grocery"),
        ("C001", 89.99m, "Electronics"),
        ("C002", 32.00m, "Grocery"),
        ("C002", 15.50m, "Health"),
        ("C003", 1250.00m, "Electronics"),
        ("C003", 450.00m, "Fashion"),
        ("C004", 12.99m, "Grocery"),
        ("C004", 8.50m, "Grocery"),
        ("C005", 675.00m, "Electronics"),
        ("C005", 320.00m, "Fashion")
    ];

    public static async Task<int> RunAsync(int phase, string? requestedModelId) => phase switch
    {
        1 => await Phase1HelloWorldAsync(requestedModelId),
        2 => await Phase2EventsAsync(requestedModelId),
        3 => await Phase3ToolsAsync(requestedModelId),
        4 => await Phase4PersistenceAsync(requestedModelId),
        5 => await Phase5McpAsync(requestedModelId),
        _ => PhaseUsage()
    };

    // Phase 1 — Hello World: a client and a session
    private static async Task<int> Phase1HelloWorldAsync(string? requestedModelId)
    {
        Console.WriteLine("== Phase 1: hello world ==\n");

        await using var client = new CopilotClient();
        await client.StartAsync();

        var modelId = await ModelPicker.PickAsync(client, requestedModelId);
        if (modelId is null)
        {
            return 1;
        }

        // Streaming emits token deltas as they arrive.
        var config = new SessionConfig
        {
            Model = modelId,
            Streaming = true
        };

        await using var session = await client.CreateSessionAsync(config);

        var done = new TaskCompletionSource();

        session.On<SessionEvent>(evt =>
        {
            switch (evt)
            {
                case AssistantMessageDeltaEvent delta:
                    Console.Write(delta.Data.DeltaContent ?? "");
                    break;
                case SessionIdleEvent:
                    done.TrySetResult();
                    break;
                case SessionErrorEvent error:
                    Console.WriteLine($"\nERROR: {error.Data.Message}");
                    done.TrySetException(new Exception(error.Data.Message));
                    break;
            }
        });

        Console.WriteLine("Prompt: Name three retail KPIs. One line each.\n");
        await session.SendAsync(new MessageOptions
        {
            Prompt = "Name three retail KPIs. One line each."
        });

        await done.Task.WaitAsync(TimeSpan.FromMinutes(3));
        Console.WriteLine();
        return 0;
    }

    // Phase 2 — Events: what the session emits
    private static async Task<int> Phase2EventsAsync(string? requestedModelId)
    {
        Console.WriteLine("== Phase 2: events ==\n");

        await using var client = new CopilotClient();
        await client.StartAsync();

        var modelId = await ModelPicker.PickAsync(client, requestedModelId);
        if (modelId is null)
        {
            return 1;
        }

        await using var session = await client.CreateSessionAsync(new SessionConfig
        {
            Model = modelId,
            Streaming = true
        });

        var done = new TaskCompletionSource();
        var order = 0;
        var deltaCount = 0;

        // Count streaming deltas so the lifecycle remains readable.
        session.On<SessionEvent>(evt =>
        {
            var name = evt.GetType().Name;

            if (name.EndsWith("DeltaEvent", StringComparison.Ordinal))
            {
                deltaCount++;
                return;
            }

            // Generic envelopes do not represent a lifecycle step.
            if (name is nameof(SessionEvent))
            {
                return;
            }

            Console.WriteLine($"{++order,3}. {name}");

            if (evt is SessionIdleEvent)
            {
                done.TrySetResult();
            }
        });
        await session.SendAsync(new MessageOptions
        {
            Prompt = "In one sentence: what is customer churn?"
        });

        await done.Task.WaitAsync(TimeSpan.FromMinutes(3));
        Console.WriteLine($"\n(+ {deltaCount} *DeltaEvent suppressed — that flood IS the streaming)");
        return 0;
    }

    // Phase 3 — Tools: descriptions tell the model when to call the tool
    [Description("Gets the total amount a given retail customer has spent.")]
    private static string GetCustomerTotal(
        [Description("Customer identifier, for example C003")] string customerId)
    {
        var matches = Transactions.Where(t =>
            string.Equals(t.CustomerId, customerId, StringComparison.OrdinalIgnoreCase)).ToArray();

        if (matches.Length == 0)
        {
            return $"No transactions found for {customerId}.";
        }

        var total = matches.Sum(t => t.Amount);

        // Visible proof that the custom tool executed.
        Console.WriteLine($"  [tool] GetCustomerTotal({customerId}) -> {total:C}");
        return $"{customerId} has {matches.Length} transactions totalling {total:C}.";
    }

    private static async Task<int> Phase3ToolsAsync(string? requestedModelId)
    {
        Console.WriteLine("== Phase 3: tools ==\n");

        await using var client = new CopilotClient();
        await client.StartAsync();

        var modelId = await ModelPicker.PickAsync(client, requestedModelId);
        if (modelId is null)
        {
            return 1;
        }

        // Run once with Tools = [] to demonstrate the CLI's built-in tools.
        // ✍️ TYPE THIS LIVE (Phase 3 only):
        //     var totalTool = CopilotTool.DefineTool(GetCustomerTotal);
        // Then replace Tools = [] with Tools = [totalTool].

        var config = new SessionConfig
        {
            Model = modelId,
            Streaming = false,
            Tools = []
        };

        await using var session = await client.CreateSessionAsync(config);

        var done = new TaskCompletionSource();

        session.On<SessionEvent>(evt =>
        {
            switch (evt)
            {
                case AssistantMessageEvent msg when !string.IsNullOrWhiteSpace(msg.Data.Content):
                    Console.WriteLine($"\nAssistant: {msg.Data.Content}");
                    break;
                case SessionIdleEvent:
                    done.TrySetResult();
                    break;
                case SessionErrorEvent error:
                    Console.WriteLine($"\nERROR: {error.Data.Message}");
                    done.TrySetException(new Exception(error.Data.Message));
                    break;
            }
        });

        // Let the model select the tool without naming it in the prompt.
        Console.WriteLine("Prompt: How much has customer C003 spent in total?\n");
        await session.SendAsync(new MessageOptions
        {
            Prompt = "How much has customer C003 spent in total?"
        });

        await done.Task.WaitAsync(TimeSpan.FromMinutes(3));
        return 0;
    }

    // Phase 4 — Persistence: dispose, then resume
    private static async Task<int> Phase4PersistenceAsync(string? requestedModelId)
    {
        Console.WriteLine("== Phase 4: persistence ==\n");

        await using var client = new CopilotClient();
        await client.StartAsync();

        var modelId = await ModelPicker.PickAsync(client, requestedModelId);
        if (modelId is null)
        {
            return 1;
        }

        // The session ID enables resume after disposal.
        var sessionId = $"codealong-{Guid.NewGuid():N}"[..24];
        Console.WriteLine($"Session id: {sessionId}\n");

        Console.WriteLine("--- Turn 1 (new session) ---");
        await using (var session = await client.CreateSessionAsync(new SessionConfig
        {
            SessionId = sessionId,
            Model = modelId,
            Streaming = false
        }))
        {
            await SendAndPrintAsync(session,
                "Remember this: my favourite retail segment is 'At Risk'. Reply with just OK.");
        }

        Console.WriteLine("\nSession disposed.\n");

        Console.WriteLine("--- Turn 2 (resumed) ---");
        await using (var resumed = await client.ResumeSessionAsync(sessionId, new ResumeSessionConfig
        {
            Model = modelId,
            Streaming = false
        }))
        {
            await SendAndPrintAsync(resumed,
                "Which retail segment did I say was my favourite?");
        }
        return 0;
    }

    // Phase 5 — MCP: attach tools from a server
    // Network dependent: use the recording if learn.microsoft.com is blocked.
    private static async Task<int> Phase5McpAsync(string? requestedModelId)
    {
        Console.WriteLine("== Phase 5: MCP ==\n");

        await using var client = new CopilotClient();
        await client.StartAsync();

        var modelId = await ModelPicker.PickAsync(client, requestedModelId);
        if (modelId is null)
        {
            return 1;
        }

        var config = new SessionConfig
        {
            Model = modelId,
            Streaming = false,
            McpServers = new Dictionary<string, McpServerConfig>
            {
                ["microsoft.docs.mcp"] = new McpHttpServerConfig
                {
                    Url = "https://learn.microsoft.com/api/mcp"
                }
            },
            OnPermissionRequest = PermissionHandler.ApproveAll
        };

        await using var session = await client.CreateSessionAsync(config);

        var done = new TaskCompletionSource();

        session.On<SessionEvent>(evt =>
        {
            switch (evt)
            {
                case ToolExecutionStartEvent start:
                    // Show whether each call came from MCP or a built-in tool.
                    Console.WriteLine(!string.IsNullOrWhiteSpace(start.Data.McpServerName)
                        ? $"  [mcp] {start.Data.McpServerName} :: {start.Data.McpToolName}"
                        : $"  [built-in] {start.Data.ToolName}");
                    break;
                case AssistantMessageEvent msg when !string.IsNullOrWhiteSpace(msg.Data.Content):
                    Console.WriteLine($"\nAssistant: {msg.Data.Content}");
                    break;
                case SessionIdleEvent:
                    done.TrySetResult();
                    break;
                case SessionErrorEvent error:
                    Console.WriteLine($"\nERROR: {error.Data.Message}");
                    done.TrySetException(new Exception(error.Data.Message));
                    break;
            }
        });

        Console.WriteLine("Prompt: Search Microsoft Learn — what is Azure Container Apps?\n");
        await session.SendAsync(new MessageOptions
        {
            // Explicit steering prevents a built-in web fetch from hiding the MCP call.
            Prompt = "Use the microsoft.docs.mcp tools to search Microsoft Learn, then tell me "
                   + "in two sentences what Azure Container Apps is."
        });

        await done.Task.WaitAsync(TimeSpan.FromMinutes(3));
        return 0;
    }

    // Helpers
    private static async Task SendAndPrintAsync(CopilotSession session, string prompt)
    {
        var done = new TaskCompletionSource();

        session.On<SessionEvent>(evt =>
        {
            switch (evt)
            {
                case AssistantMessageEvent msg when !string.IsNullOrWhiteSpace(msg.Data.Content):
                    Console.WriteLine($"Assistant: {msg.Data.Content}");
                    break;
                case SessionIdleEvent:
                    done.TrySetResult();
                    break;
                case SessionErrorEvent error:
                    Console.WriteLine($"ERROR: {error.Data.Message}");
                    done.TrySetException(new Exception(error.Data.Message));
                    break;
            }
        });

        Console.WriteLine($"You: {prompt}");
        await session.SendAsync(new MessageOptions { Prompt = prompt });
        await done.Task.WaitAsync(TimeSpan.FromMinutes(3));
    }

    private static int PhaseUsage()
    {
        Console.WriteLine("""
            Usage:
              dotnet run --project src/AgentOrchestrator/samples/SdkLabs -- codealong --phase <1-5>

              1  Hello world     — client + session, streaming
              2  Events          — what the session emits, in order
              3  Tools           ⭐ define a tool the model calls
              4  Persistence     — dispose, then resume the same session
              5  MCP             — attach a server you didn't write (needs network)
            """);
        return 1;
    }
}
